server/src/main/java/org/elasticsearch/cluster/health/ClusterHealthStatus.java

```diff
@@ -64,4 +67,19 @@ public static ClusterHealthStatus fromString(String status) {
             throw new IllegalArgumentException("unknown cluster health status [" + status + "]");
         }
     }
+
+    public static ClusterHealthStatus fromClusterState(ClusterState clusterState, Index index) {
+        IndexRoutingTable indexRoutingTable = clusterState.routingTable().index(index);
+        if (indexRoutingTable == null || indexRoutingTable.allPrimaryShardsActive() == false) {
+            return ClusterHealthStatus.RED;
+        }
```

Comment:
I think this is different than the definition from the constructor of `ClusterIndexHealth` (where the `status` is calculated). The reason is that `indexRoutingTable.allPrimaryShardsActive()` will return "true" if there are 0 shards, which would normally be a "RED" state:

 https://github.com/elastic/elasticsearch/blob/3027039f72c26f6780cd7857a6c5c4b0c4bd0878/server/src/main/java/org/elasticsearch/cluster/health/ClusterIndexHealth.java#L151-L153

But here, we'd return green in that case, because the subsequent `for` loop would loop over 0 shards and the final `return` statement would return.

I think, if possible, we should try to re-use the logic from `ClusterIndexHealth` so that we calculate the health color in as few places as possible. If that's not possible, we need to change this to return "red" if there are 0 shards so it matches the existing behavior.

Reply:
I completely agree, we should reuse it, I stumbled across this logic from `WaitForIndexColorStep` first and I stopped looking. Changed.